File: naver_blog_image_downloader/ios/Runner/Applications/Channels/Features/AppIconChannel.swift
import Flutter
import UIKit

// MARK: - App Icon Channel

extension AppDelegate {

    /// 註冊 AppIcon MethodChannel，處理 `setAppIcon` 與 `getCurrentIcon` 呼叫。
    ///
    /// - Parameter messenger: Flutter BinaryMessenger，用於建立 MethodChannel。
    func setupAppIconChannel(messenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(
            name: "com.leoho.naverBlogImageDownloader/appIcon",
            binaryMessenger: messenger
        )

        channel.setMethodCallHandler { [weak self] call, result in
            guard let self else {
                result(FlutterError(
                    code: "UNAVAILABLE",
                    message: "AppDelegate deallocated",
                    details: nil
                ))
                return
            }
            self.handleAppIconMethodCall(call, result: result)
        }
    }
}

// MARK: - Private Methods

private extension AppDelegate {

    /// 分派 AppIcon MethodChannel 的方法呼叫。
    ///
    /// - Parameters:
    ///   - call: Flutter 端傳入的方法呼叫，包含方法名稱與參數。
    ///   - result: 回傳結果給 Flutter 端的 callback。
    func handleAppIconMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            switch call.method {
            case "setAppIcon":
                await handleSetAppIcon(call, result: result)
            case "getCurrentIcon":
                handleGetCurrentIcon(result: result)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    /// 處理 `setAppIcon` 方法呼叫，透過 UIApplication 切換 App 圖示。
    ///
    /// 使用 xcassets icon set 方式：
    /// `"default"` 對應 `nil`（主圖示），`"new"` 對應 `"NewAppIcon"`（替代圖示 icon set 名稱）。
    ///
    /// - Parameters:
    ///   - call: 需包含 `Map` 類型的參數，含 `iconName`（String）。
    ///   - result: 成功回傳 `nil`，失敗回傳 `FlutterError`。
    func handleSetAppIcon(_ call: FlutterMethodCall, result: @escaping FlutterResult) async {
        guard let args = call.arguments as? [String: Any],
              let iconName = args["iconName"] as? String else {
            result(FlutterError(code: "INVALID_ARG", message: "iconName is required", details: nil))
            return
        }

        let alternateIconName: String? = (iconName == "default") ? nil : "NewAppIcon"

        do {
            try await UIApplication.shared.setAlternateIconName(alternateIconName)
            result(nil)
        } catch {
            result(FlutterError(code: "SET_ICON_FAILED", message: error.localizedDescription, details: nil))
        }
    }

    /// 處理 `getCurrentIcon` 方法呼叫，取得目前使用中的 App 圖示名稱。
    ///
    /// `alternateIconName` 為 `nil` 表示使用主圖示（classic），非 `nil` 表示使用替代圖示（new）。
    ///
    /// - Parameter result: 回傳 `"default"` 或 `"new"`。
    func handleGetCurrentIcon(result: @escaping FlutterResult) {
        let currentAlternate = UIApplication.shared.alternateIconName
        let iconName = (currentAlternate == nil) ? "default" : "new"
        result(iconName)
    }
}

File: naver_blog_image_downloader/android/app/src/main/kotlin/com/leoho/naverBlogImageDownloader/android/applications/channels/features/AppIconChannel.kt
package com.leoho.naverBlogImageDownloader.android.applications.channels.features

import android.content.ComponentName
import android.content.pm.PackageManager
import com.leoho.naverBlogImageDownloader.android.applications.MainActivity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

// region Channel Setup

/** AppIcon MethodChannel 通道名稱。 */
private const val APP_ICON_CHANNEL_NAME = "com.leoho.naverBlogImageDownloader/appIcon"

/** 預設圖示的 activity-alias 完整 component 名稱。 */
private const val ALIAS_CLASSIC = "com.leoho.naverBlogImageDownloader.android.MainActivityClassic"

/** 新版圖示的 activity-alias 完整 component 名稱。 */
private const val ALIAS_NEW = "com.leoho.naverBlogImageDownloader.android.MainActivityNew"

/**
 * 註冊 AppIcon MethodChannel，處理 `setAppIcon` 與 `getCurrentIcon` 呼叫。
 *
 * @param flutterEngine Flutter 引擎實例，用於取得 BinaryMessenger。
 */
fun MainActivity.setupAppIconChannel(flutterEngine: FlutterEngine) {
    MethodChannel(flutterEngine.dartExecutor.binaryMessenger, APP_ICON_CHANNEL_NAME)
        .setMethodCallHandler { call, result ->
            handleAppIconMethodCall(call, result)
        }
}

// endregion

// region Private Methods

/**
 * 分派 AppIcon MethodChannel 的方法呼叫。
 *
 * @param call Flutter 端傳入的方法呼叫，包含方法名稱與參數。
 * @param result 回傳結果給 Flutter 端的 callback。
 */
private fun MainActivity.handleAppIconMethodCall(
    call: MethodCall,
    result: MethodChannel.Result
) {
    when (call.method) {
        "setAppIcon" -> handleSetAppIcon(call, result)
        "getCurrentIcon" -> handleGetCurrentIcon(result)
        else -> result.notImplemented()
    }
}

/**
 * 處理 `setAppIcon` 方法呼叫，透過 PackageManager 啟用/停用 activity-alias 切換圖示。
 *
 * 使用 `DONT_KILL_APP` 旗標避免切換時殺掉 App。
 *
 * @param call 需包含 `Map` 類型的參數，含 `iconName`（String，`"default"` 或 `"new"`）。
 * @param result 成功回傳 null，失敗回傳錯誤。
 */
private fun MainActivity.handleSetAppIcon(
    call: MethodCall,
    result: MethodChannel.Result
) {
    val args = call.arguments as? Map<*, *>
    val iconName = args?.get("iconName") as? String
    if (iconName == null) {
        result.error("INVALID_ARG", "iconName is required", null)
        return
    }

    try {
        val pm = packageManager
        val enableAlias: String
        val disableAlias: String

        if (iconName == "default") {
            enableAlias = ALIAS_CLASSIC
            disableAlias = ALIAS_NEW
        } else {
            enableAlias = ALIAS_NEW
            disableAlias = ALIAS_CLASSIC
        }

        pm.setComponentEnabledSetting(
            ComponentName(this, disableAlias),
            PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
            PackageManager.DONT_KILL_APP
        )
        pm.setComponentEnabledSetting(
            ComponentName(this, enableAlias),
            PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
            PackageManager.DONT_KILL_APP
        )
        result.success(null)
    } catch (e: Exception) {
        result.error("SET_ICON_FAILED", e.message, null)
    }
}

/**
 * 處理 `getCurrentIcon` 方法呼叫，檢查哪個 activity-alias 目前啟用中。
 *
 * @param result 回傳 `"default"` 或 `"new"`。
 */
private fun MainActivity.handleGetCurrentIcon(result: MethodChannel.Result) {
    val pm = packageManager
    val newState = pm.getComponentEnabledSetting(
        ComponentName(this, ALIAS_NEW)
    )
    val iconName = if (newState == PackageManager.COMPONENT_ENABLED_STATE_ENABLED) {
        "new"
    } else {
        "default"
    }
    result.success(iconName)
}

// endregion
